README.md
# rsc - A generic RightScale API client

[![License](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/rightscale/rsc/blob/master/LICENSE) [![Godoc](https://godoc.org/github.com/rightscale/rsc?status.svg)](http://godoc.org/github.com/rightscale/rsc)

Master
[![Build Status](https://travis-ci.org/rightscale/rsc.svg?branch=master)](https://travis-ci.org/rightscale/rsc)

v9.0.0
[![Build Status](https://travis-ci.org/rightscale/rsc.svg?branch=v9.0.0)](https://travis-ci.org/rightscale/rsc)

`rsc` provides both a command line tool and a go package for interacting with the RightScale APIs.
The currently supported APIs are the RightScale Cloud Management API 1.5 and 1.6 APIs, the
RightScale Self-Service 1.0 APIs (latest version for this product) and the RightLink10 APIs exposed
by the RightLink10 agent. The RightScale APIs reference can be found on the
[RightScale docs](http://docs.rightscale.com/api/) site.

`rsc` can be used in one of two ways:

- as a command line tool
- as a way to make API requests to RightScale programmatically from Go code

## <a name="usage"></a>Command Line Tool

The command line tool uses subcommands to interact with each API. Use `rsc cm15` to send requests
to the RightScale Cloud Management API 1.5, `rsc cm16` to send requests to the RightScale Cloud
Management API 1.6, `rsc ss` to send requests to the RightScale Self-Service API 1.0 and `rsc rl10`
to send requests to RightLink10.

### Installation

The `rsc` command line tool is a statically linked binary making installation a breeze. There is
no dependency on any runtime library. Just download the correct version for your OS and
architecture and you're good to go.

The latest stable versions can be download from:

- macOS Intel: `https://binaries.rightscale.com/rsbin/rsc/v9/rsc-darwin-amd64.tgz`
- macOS ARM64: `https://binaries.rightscale.com/rsbin/rsc/v9/rsc-darwin-arm64.tgz`
- Windows: `https://binaries.rightscale.com/rsbin/rsc/v9/rsc-windows-amd64.zip`
- Linux: `https://binaries.rightscale.com/rsbin/rsc/v9/rsc-linux-amd64.tgz`
- Linux ODroid/RasPi/armhf: `https://binaries.rightscale.com/rsbin/rsc/v9/rsc-linux-arm.tgz`

As an example the following downloads and runs the macOS version:

```bash
$ curl https://binaries.rightscale.com/rsbin/rsc/v9/rsc-darwin-amd64.tgz | tar -zxf - -O rsc/rsc > rsc
$ chmod +x ./rsc
$ ./rsc --version
rsc v9.0.0 - 2021-04-13 22:01:22 - 4d8e4ad2a7f54fd085cb267df6c0baa60de0ad53
```

#### Versioning

- To download the latest stable use the links with 'v9' in them.
- To download a specific version, replace the 'v9' by the exact version, such as 'v9.0.0'.
- All versions with the same major number (e.g. 'v9') are intended to be "upward" compatible.
- The 'v9' links download a specific version, so `rsc --version` will print something like 'v9.0.0'
  and not 'v9'.
- The latest dev version is 'master'.

### Command Line

The general shape of a command line is:

```text
rsc [GLOBAL] [API] ACTION HREF [PARAM=VALUE]
```

where:

- `GLOBAL` is an optional list of global flags,
- `API` is `cm15`, `cm16`, `ss` or `rl10`
- `ACTION` is the API action to perform (i.e `index`, `show`, `update`, etc.),
- `HREF` is the resource or resource collection href (i.e. `/api/servers`, `/api/servers/1` etc.), and
- `PARAM` and `VALUE` are the names and values of the action parameters (e.g. `view=extended`).

For example:

```bash
rsc cm15 show /api/servers 'filter[]=name==LB'
```

The sections below cover each option in order.

### Global Flags

The list of global flags is:

```text
  --help           Show help.
  --version        Show application version.
  -c, --config="/home/raphael/.rsc"
                   path to rsc config file
  -R, --retry=0    Number of retry attempts for non-successful API responses (500, 503, and timeouts only)
  -a, --account=ACCOUNT
                   RightScale account ID
  -h, --host=HOST  RightScale login endpoint (e.g. 'us-3.rightscale.com')
  --email=EMAIL    Login email, use --email and --pwd or use --refreshToken, --accessToken, --apiToken or --rl10
  --pwd=PWD        Login password, use --email and --pwd or use --refreshToken, --accessToken, --apiToken or --rl10
  -r, --refreshToken=REFRESHTOKEN
                   OAuth refresh token, use --email and --pwd or use --refreshToken, --accessToken, --apiToken or --rl10
  -s, --accessToken=ACCESSTOKEN
                   OAuth access token, use --email and --pwd or use --refreshToken, --accessToken, --apiToken or --rl10
  -p, --apiToken=APITOKEN
                   Instance API token, use --email and --pwd or use --refreshToken, --accessToken, --apiToken or --rl10
  -F, --flexeraOne Use Flexera One for authentication, with --refreshToken
  --rl10           Proxy requests through RightLink 10 agent, use --email and --pwd or use --refreshToken, --accessToken, --apiToken or --rl10
  --noAuth         Make unauthenticated requests, used for testing
  --x1=X1          Extract single value using JSON:select
  --xm=XM          Extract zero, one or more values using JSON:select and return newline separated list
  --xj=XJ          Extract zero, one or more values using JSON:select and return JSON
  --xh=XH          Extract header with given name
  --fetch          Fetch resource with href present in 'Location' header
  --dump=DUMP      Dump HTTP request and response. Possible values are 'debug' or 'json'.
  -v, --verbose    Dump HTTP request and response including auth requests and headers, enables --dump=debug by default, use --dump=json to switch format
  --pp             Pretty print response body
```

### Authentication

There are 3 different mechanisms for authenticating with the RightScale platform.

#### Basic Authentication

Basic authentication uses an email and password to create a session against the Cloud Management
APIs (a.k.a. API 1.5). Creating a session returns a cookie. The session cookie can be used to make
authenticated requests against all accounts the authenticated user has access to. While the session
is created by making an API request to the CM 1.5 APIs the resulting cookie can be used to make API
calls against all the RightScale APIs enabled for the user.

`rsc` supports basic authentication via the `--email` and `--pwd` flags. When using `rsc` as a
Go package authentication is done once and the same cookie is used for all API requests made with
the same client. The package also takes care of refreshing the cookie before the session expires.

Below is an example listing all clouds available in a given account using the `rsc` command line
tool with basic authentication:

```bash
rsc --account $ACCOUNT --email $EMAIL --pwd $PASSWORD --host $HOST cm15 index clouds
```

The example assumes that the ACCOUNT, EMAIL, PASSWORD and HOST environment variables contain the
account id, user email, password and RightScale API host respectively.

At the time of writing there are two possible values for the RightScale API host:
`us-3.rightscale.com` or `us-4.rightscale.com`. When using basic authentication it is also
possible to use `my.rightscale.com` in which case the server will redirect to the appropriate
host. The correct value for the host can be retrieved by logging in into the RightScale Cloud
Management dashboard: the URL contains the appropriate host for the account being logged into.

#### OAuth Authentication

RightScale supports the [OAuth 2.0 Authorization Framework](http://tools.ietf.org/html/rfc6749)
where a refresh token can be exchanged with a temporary access token to make authenticated requests.
The refresh token is created using the RightScale Cloud Management dashboard from the *Settings >
Account Settings > API Credentials* menu. Access tokens are then created using the
[`OAuth2`](http://reference.rightscale.com/api1.5/resources/ResourceOauth2.html#create) resource
exposed by the Cloud Management 1.5 APIs. Note that refresh and access tokens are
*account specific*. This has two consequences: refresh tokens must be retrieved for each account and
it is not necessary to specify the account when making an API request with an access token.

`rsc` supports OAuth authentication via the `--refreshToken` and `--accessToken` flags. If the
refresh token is used then `rsc` takes care of creating an access token and uses the access token
to make the final API request. This results in two API requests which may not be optimal in certain
scenarios (e.g. scripts calling `rsc` to make multiple API requests). For these scenarios `rsc` can
also use a pre-existing access token directly. That access token can be created using the OAuth2
resource.

Below is an example listing all clouds available in the account using the refresh token retrieved
from the RightScale Cloud Management dashboard:

```bash
rsc --refreshToken $REFRESH --host $HOST cm15 index clouds
```

Note that in this case the account doesn't need to be specified on the command line, it is
inferred from the token.

Here is another example that first creates an access token explicitly then uses that token to list
all clouds:

```bash
export ACCESS=`rsc --x1 .access_token --refreshToken $REFRESH cm15 create oauth2 grant_type=refresh_token refresh_token=$REFRESH`
rsc --accessToken $ACCESS --host $HOST cm15 index clouds
```

The example above uses the `--x1` flag to extract the access token from the response. Extracting
data from responses is described in the [Extracting values from responses](#extract) section below.

#### Instance Facing APIs

The final mechanism for authenticating against the RightScale APIs consists of using an instance
specific token to make API requests from a RightScale managed instance. This token is written to
the instance user data on launch by the RightScale platform. There are actually two different ways
scripts running on RightScale managed instances can authenticate:

- The instance API token can be used to create a session which returns a cookie. This method is
  similar to basic authentication except that a token is used instead of email and password.
- [RightLink 10](http://docs.rightscale.com/rl/about.html) also runs a HTTP proxy that can be used
  to make authenticated requests without requiring any credentials from the client.

Here is an example using the instance API token to list all clouds:

```bash
rsc --apiToken $TOKEN --host $HOST cm15 index clouds
```

And here is another example running on a RightLink 10 enabled instance:

```bash
rsc --rl10 cm15 index clouds
```

#### Storing Client Credentials

The `setup` command can be used to create a configuration file that contains the host, account id,
user email and password so that these don't need to be specified each time. All these settings or
a subset may be stored (i.e. the password doesn't have to be stored if that's not desirable).

By default the config file is created in `$HOME/.rsc`, the location can be overridden using the
`--config` global flag. Multiple configs may be created to allow for different environments or
users. Use the `--config` flag when invoking the tool to specify the location of the config file if it's
not the default.

The configuration file is a simple JSON file that lists the fields defined during setup. The
password is encrypted before being stored although it is a two way encryption scheme and so is not
meant to be a truly secure mechanism but rather a way to avoid having the password stored in plain
text on disk.

```bash
$ rsc setup
Account id: 12345
Login email: myemail@mycompany.com
Login password: 12345abc
API Login host: us-3.rightscale.com
```

The values stored in the configuration can be overridden using command line flags so that for
example a different account can be specified:

```bash
rsc --account 234 cm15 index clouds
```

### <a name=extract></a>Extracting Values From Responses

The `--x1`, `--xm` and `--xj` flags make it possible to extract values from the response using a
JSON select expression (see [https://github.com/lloyd/JSONSelect](https://github.com/lloyd/JSONSelect)). For example:

```bash
rsc --xm .name cm15 index /api/clouds
```

extracts the names of each cloud from the response and prints the result as a newline separated list
which is convenient to consume from bash scripts:

```bash
$ declare -a "clouds=(`rsc --xm=.cloud_type cm15 index clouds`)"; set | egrep '^clouds'
clouds=([0]="amazon" [1]="open_stack_v2" [2]="cloud_stack" [3]="rackspace_next_gen" [4]="google" [5]="azure" [6]="soft_layer" [7]="vscale")
```

For additional help on extracting values see the [Command Line Help and Cookbook](COOKBOOK.md).

### Actions and Parameters

The names of the actions available for a given API or a given API resource can be listed with the
special `actions` action:

```bash
rsc cm15 actions
```

(output of example above not shown for brevity)

To get the actions available on a resource specify the resource href as in:

```bash
$ rsc cm15 actions /api/clouds/1/volumes
Action  Href                              Resource
======= ================================= ========
create  /api/clouds/:cloud_id             Volume
------- --------------------------------- --------
destroy /api/clouds/:cloud_id/volumes/:id Volume
------- --------------------------------- --------
index   /api/clouds/:cloud_id             Volume
------- --------------------------------- --------
show    /api/clouds/:cloud_id/volumes/:id Volume
```

Parameters use URL form encoding to represent nested data structures used in request bodies. For example
using the RightScale CM API 1.5 the command line to create a volume is:

```bash
$ rsc --pp --fetch cm15 create /api/clouds/1/volumes \
  'volume[name]=My New Volume' \
  'volume[size]=10' \
  'volume[datacenter_href]=/api/clouds/1/datacenters/5K443K2CF8NS6' \
  'volume[volume_type_href]=/api/clouds/1/volume_types/BDVEN383N1EN2'
```

The `--pp` and `--fetch` options above are optional, `--fetch` makes a subsequent API call to retrieve
the newly created resource and `--pp` pretty prints the response.
Use the name of the fields followed by `[]` to represent arrays:

```bash
$ rsc cm16 index /api/deployments 'filter[]=description==awesome deployment' \
  'filter[]=name==app servers'
```

The `/api/` prefix for CM API 1.5 and CM API 1.6 hrefs is optional so the following lists all
deployments:

```bash
rsc cm16 index deployments
```

### Built-in Help

The `--help` flag is available on all commands. It displays contextual help, for example:

```text
$ rsc index --help
usage: rsc [<flags>] index <href> [<params>]

Lists all resources of given type in account.

Args:
  <href>      API Resource or resource collection href on which to act, e.g. '/api/servers'
  [<params>]  Action parameters in the form QUERY=VALUE, e.g. 'server[name]=server42'
```

Or:

```text
$ rsc cm15 index clouds --help
usage: rsc [<flags>] index index /api/clouds [<cloud index params>]

Cloud index params:
filter[]=[]string
    <optional, [cloud_type|description|name]>

view=string
    <optional, [default|extended]>
```

The help lists the valid values for views and filters for example.
It also indicates which flags are mandatory.

-----

## <a name="go"></a>Go Package

The other use case for `rsc` is making programmatic API requests to the
RightScale platform. Each API client code is encapsulated in a different
sub-package: package `cm15` for CM API 1.5, package `cm16`for CM API
1.6, package `ss` for Self-service APIs and package `rl10` for RightLink10 requests.

### Installation

`rsc` uses gopkg.in for versioning, this means that you can download the released `rsc` packages
as follows:

```bash
go get gopkg.in/rightscale/rsc.v9
```

and import then in your code with:

```go
import "gopkg.in/rightscale/rsc.v9"
```

If you intend on contributing, just want to play around with the code or feel adventurous you can
download and use the bleeding edge version from github which corresponds to the master branch:

```bash
go get github.com/rightscale/rsc
```

```go
import "github.com/rightscale/rsc
```

### Client Creation

Each API client package defines an `API` struct that represents the API
client. Clients are created using one of three factory methods: `New`,
`NewRL10` or `FromCommandLine`. The latter is used by the top level
`main` package to create clients from the values provided on the command
line. `NewRL10` is meant to be called by code that runs on a RightScale
server running the RightLink10 agent. It configures the client to
talk to the APIs through the proxy exposed by RightLink10. Overall
the most flexible function is `New` which accepts an API host name,
a RightScale account ID, authentication information and an optional
low-level HTTP client. As an example the following creates a
CM API 1.5 client that connects to `us-3.rightscale.com` using a OAuth
refresh token for authentication and the default HTTP client:

```go
// Retrieve refresh tokens from the RightScale dashboard Settings/API Credentials menu
refreshToken := "3e040efed9a83ac758f3b1cbdfa041b905742169"
// Corresponding RightScale account ID
accountID := 60073
auth := rsapi.NewOAuthAuthenticator(refreshToken, accountID)
accountId := 123
client := cm15.New("us-3.rightscale.com", &auth, nil)
```

You may test the credentials using the `CanAuthenticate` method:

```go
if err := client.CanAuthenticate(); err != nil {
    // Woops creds are not working
    return err
}
```

This method makes a test API request so is expensive, the idea is to call it once then use the client
to make a series of requests.

### Logging

The `log` package exposes a `Logger` variable of type `log15.Logger`. This logger is used
by rsc to log API requests made to external services. Configure the logger handler
to enable logging. The [log15](https://godoc.org/github.com/inconshreveable/log15) package comes
with a number of default handlers including a file and a syslog handlers.

Configuring the `log` package to log to the file `/var/log/myapp.log` would look like:

```go
handler, err := log15.FileHandler("/var/log/myapp.log", log15.LogfmtFormat())
if err != nil {
    return err
}
log.Logger.SetHandler(handler)
```

Consult the [log15](https://godoc.org/github.com/inconshreveable/log15) GoDocs for more information.

### Locators

Once a client has been created resources can be retrieved using resource locators. The code defines one
resource locator type per resource exposed by the underlying API.

Locators are instantiated using factory methods exposed by the client object. The factory methods
accept the collection or resource href and return the corresponding locator. For example the
following creates a cloud locator:

```go
var cloudLocator = client.CloudLocator("/api/clouds/1")
```

Locators expose one method for each action supported by the underlying collection or resource. For
example the clouds collection locator `CloudLocator` exposes an `Index()` and a `Show()` method.
Locator methods may return resources which are structs that expose the underlying resource
attributes. For example the `CloudLocator` `Index()` method returns an array of `Cloud` resource.
A cloud resource is defined as:

```go
// Represents a Cloud (within the context of the account in the session).
type Cloud struct {
    Capabilities []map[string]interface{} `json:"capabilities,omitempty"`
    CloudType    string                   `json:"cloud_type,omitempty"`
    Description  string                   `json:"description,omitempty"`
    DisplayName  string                   `json:"display_name,omitempty"`
    Links        []map[string]string      `json:"links,omitempty"`
    Name         string                   `json:"name,omitempty"`
}
```

and the `Index()` method is defined as:

```go
// GET /api/clouds
// Lists the clouds available to this account.
// Optional parameters:
// filter
// view
func (loc *CloudLocator) Index(options rsapi.APIParams) ([]*Cloud, error)
```

The following code would invoke the `Index()` method using the default view and no filter to make the API request:

```go
var clouds, err = api.CloudLocator("/api/clouds").Index(rsapi.APIParams{})
```

`Create` actions all return a locator so that fetching the corresponding resource is easy:

```go
var volumesLocator = client.VolumeLocator("/api/clouds/1/volumes")
var params cm15.VolumeParam{} // Code that sets parameters omitted for brevity
loc, err := volumeLocator.Create(&params)
if err == nil {
    volume, err := loc.Show(rsapi.APIParams{})
    // ... check error, use volume etc.
}
```

It is also possible to create a locator directly from a resource by using the resource `Locator`
method:

```go
clouds, err := client.CloudLocator("/api/clouds").Index()
if err == nil {
    first := clouds[0].Locator(client)
    first.Show() // first is a CloudLocator instance
}
```

### Using the Generic Methods

So far we've seen how you can interact with the APIs using strongly
typed methods which are handy if you need to make specific API calls
from your code. However these methods don't work well if you need to
write a generic client that may need to make any arbitrary API call
given the names of a resource and an action and generic parameters.

For this use case each API client package also contains a generic `BuildRequest`
method which accepts the name of a resource, the name of an action,
the href of the resource and a map of generic parameters (in the form of
`map[string]interface{}`):

```go
func (a *API) BuildRequest(resource, action, href string, params rsapi.APIParams) (*http.Request, error)
```

The client also exposes a `PerformRequest` method that makes the request and optionally
dumps the request body and response to STDERR for debugging:

```go
func (a *API) PerformRequest(req *http.Request) (*http.Response, error)
```

The `httpclient` package exposes a `DumpFormat` variable that controls how much logging is done
when HTTP requests are done. The default consists of logging the request method and URL as well
as the response code and timing information. Setting `DumpFormat` to `httpclient.Debug` causes
the request and response bodies to get logged as well using the Debug log level.

### Common code

The package `rsapi` contains common code for all client packages. It
also defines an API struct that each client embeds as an anonymous field
and leverages for all common code. One such method that may be of use in
your code is `LoadResponse` that simply unmarshals the response body JSON
and returns the result. If the response contains a `Location` header (all
`create` actions return one) then the function returns a map containing
the value of the location under the `"Location"` key. The signature of
`LoadResponse` is:

```go
func (a *API) LoadResponse(resp *http.Response) (interface{}, error)
```

The `rsapi` package also includes authenticators which signs API requests
by adding the required auth headers (cookie in the case of email/password
authentication, OAuth header in the case of OAuth and custom header in
the case of RightLink10). Finally it contains common code used by all
the clients to parse the command line.

### Examples

The repository includes the following Go examples:

- [Cloud Management 1.5 Basic Example](https://github.com/rightscale/rsc/tree/master/cm15/examples/basic)
- [Cloud Management 1.5 Auditail Example](https://github.com/rightscale/rsc/tree/master/cm15/examples/auditail)
- [Cloud Management 1.5 RS SSH Example](https://github.com/rightscale/rsc/tree/master/cm15/examples/rsssh)
- [Self-Service Basic Example](https://github.com/rightscale/rsc/tree/master/ss/examples/basic)

### Reference Documentation

The documentation for each package is hosted on godoc.org:

- [Cloud Management 1.5](http://godoc.org/github.com/rightscale/rsc/cm15)
- [Cloud Management 1.6](http://godoc.org/github.com/rightscale/rsc/cm16)
- [Self-Service Designer](http://godoc.org/github.com/rightscale/rsc/ss/ssd)
- [Self-Service Catalog](http://godoc.org/github.com/rightscale/rsc/ss/ssc)
- [Self-Service Manager](http://godoc.org/github.com/rightscale/rsc/ss/ssm)
- [RightLink10](http://godoc.org/github.com/rightscale/rsc/rl10)

-----

## <a name="contributing"></a>Development & Contributing

### Building

The following make targets are useful:

- `make depend` installs required tools
- `make` builds a binary for your local OS
- `make build` builds binaries for Linux, OS-X and Windows
- `make test` runs the test suite

#### Your own build of the latest release version

The simple option is `go get gopkg.in/rightscale/rsc.v9`, this will use the checked-in
code-generated files.

The more involved option is:

```bash
mkdir -p $GOPATH/src/gopkg.in/rightscale
cd $GOPATH/src/gopkg.in/rightscale
git clone https://github.com/rightscale/rsc.git rsc.v9
cd rsc.v9
git checkout v9.0.0
make depend
make
```

### Code generation

Part of the `rsc` source code (the vast majority in terms of lines of code) is automatically
generated from API metadata. There are currently two code generators: `api15gen` consumes the
RightScale CM API 1.5 metadata hosted [here](http://reference.rightscale.com/api1.5/api_data.json) and
`praxisgen` consumes the metadata for any [praxis](http://praxis-framework.io/) application (for example
for the RightScale CM API 1.6). `goav2gen` generates a client from a swagger generated by a
[goa v2](https://github.com/goadesign/goa/tree/v2) based design.

The source code for the code generator tools lives under the `gen` directory.
Once the tools are compiled and installed they can be invoked using `go generate`,
see [http://blog.golang.org/generate](http://blog.golang.org/generate)
for information on how `go generate` works. The `go generate` comments live in the top level file
`generate.go`.


CHANGELOG.md
ChangeLog
=========

v9.1.0 / 2022-01-25
-------------------

* Build for macOS ARM64.
* Build with latest Go and dependencies.

v9.0.0 / 2021-04-13
-------------------

* Add support for using Flexera One refresh tokens for authentication.

v8.1.0 / 2020-06-18
-------------------

* Add support for Credentials.
* Add support for selectable actions in Policies.
* Use Go modules instead of Glide.

v8.0.0 / 2020-03-19
-------------------

* Make the parsing of `RubyTime` for `cm15` more predictable by using [`time.ParseInLocation`](https://golang.org/pkg/time/#ParseInLocation) with `time.UTC`
* Build with Go 1.14.x.

v7.0.1 / 2019-11-15
-------------------

* Refreshed policy API actions
* Added ability to pass in raw json for certain parameters by doing a param:=value instead of param=value

v7.0.0 / 2019-06-07
-------------------

* All actions from API 1.5 which return the `"Location"` header now return Locator pointers from their Go functions.
* Build with Go 1.12.x.

v6.11.0 / 2019-01-08
-------------------

* add policy API to rsc (rsc policy action to see details)

v6.10.1 / 2018-08-24
--------------------

* Update the metadata to reflect the proper location for query params in index calls.

v6.10.0 / 2018-08-24
--------------------

* Support with\_deleted parameter in instance\_types show and index calls.

v6.9.0 / 2018-05-30
-------------------

* add visibility filter to VolumeAttachments index call.
* update visibility filter for Images index call.
* add scale\_up and scale\_down actions to ServerArray.

v6.8.0 / 2018-03-28
-------------------

* add api\_behavior to VolumeAttachments create call.

v6.7.0 / 2018-03-15
-------------------

* Add skip\_deletion flag for Backups.cleanup action
* Update Server resource: change deployment\_href on existing server

v6.6.0 / 2017-10-25
-------------------

* Add action/priority to security group rules.
* Add source/destination cidr IPs/group names and source start/end port to security group rules. These are currently
  only available for AzureRM.
* Add cloud\_specific\_params[service\_account] to instances

v6.5.0 / 2017-07-19
-------------------

* Add support for AzureRM Managed Disk.
* Fix a bug command line resource href matching where actions whose href pattern partially matched the href were
  considered a full match resulting in the wrong resource action being selected.

v6.4.1 / 2017-06-20
-------------------

* Fix bug introduced with storing refresh token in `~/.rsc` where it took precedence over the `-r`/`--refresh-token`
  flag rather than the flag taking precedence over the config file value like it should.

v6.4.0 / 2017-06-14
-------------------

* Build with Go 1.8 which fixes JSON output to prefer decimal notation rather than scientific notation which was messing
  up some IDs (see [encoding/json](https://golang.org/doc/go1.8#encoding\_json) in the Go 1.8 Release Notes).
* Add `httpclient.NewPB` constructor which will create an `HTTPClient` from a `ParamBlock` rather than using package
  global variables which can be unsafe with multiple Goroutines.
* Use `context` instead of `golang.org/x/net/context`.
* Support refresh token in `~/.rsc` configuration file for the command line tool.

v6.3.0 / 2017-03-02
------------------

* Added ResourceGroup resource
* Added MultiCloudImageMatcher resource
* Updated fields for Subnet, Instance, ServerArray

v6.2.0 / 2016-08-29
-------------------

* rl10: Added /rll/proc/log\_level.
* Add destroy action for alerts.
* Fields for Instance are updated.

v6.1.0 / 2016-08-09
--------

* SelfService action paths have been updated to the latest version of the SelfService API. This means that each path has been prefixed with one of "/api/catalog", "/api/design", or "/api/manager" depending on the service it is associated with. The built-in documentation has been updated with the new paths.
* The old SelfService action paths are now deprecated. While they still work as before they will no longer be supported in the next major release of RSC.
* Added support for the EndUser resource in the SelfService Catalog API.

v6.0.0 / 2016-06-28
--------

* Fields for Instance, Route, SshKey, Volume, User, and Permission types are updated.
* Instances#set\_custom\_lodgement is removed.
* InstanceCustomLodgement is removed.
* Added RightScript#delete, Servers#disable\_runnable\_bindings, Servers#enable\_runnable\_bindings, ServerArrays#disable\_runnable\_bindings, and ServerArrays#enable\_runnable\_bindings, and VolumeSnapshots#copy.
* Sessions#index now accepts a view argument and the supported views are: default and whoami.
* Parameters for Instances#create, Instances#update, Routes#create, Servers#create, Servers#update, ServerArrays#create, and Volumes#create updated.
* rl10: Add managed login actions

v5.0.3 / 2016-04-21
-------------------

* Add --retry flag. Specifies number of retry attempts for non-successful API responses(500, 503, and timeouts only)
* cm15: Add RightScriptAttachment actions. Add RightScript delete and update\_source actions.
* rl10: Add Docker integration actions.

v5.0.2 / 2016-02-05
-------------------

* Using TokenAuthenticator now sends the account ID in the X-Account header for all requests
* Handle Unicode byte order mark in JSON input since Windows PowerShell variables can include them
* Do not use a timeout when the request is associated with a context
* Support new custom monitoring plugin actions in RL10 API
* Increase the default timeout for the rsc command line client to 5 minutes (300 seconds)

v5.0.1 / 2016-01-07
-------------------

* Update gomega to fix HasOccurred
* Fix ambiguous handling of some command line arguments
* Add --timeout flag
* Add link information to resource metadata

v5.0.0 / 2015-10-15
-------------------

* Fix issue with instance session auth
* Fix optional command line parameters being required
* Properly generate code and metadata for "aliased" CM 1.5 resource actions

main.go
package main

import (
	"bytes"
	"fmt"
	"io/ioutil"
	"net"
	"net/http"
	"os"
	"strings"
	"time"

	"gopkg.in/alecthomas/kingpin.v2"
	"gopkg.in/rightscale/rsc.v9/cmd"
	"gopkg.in/rightscale/rsc.v9/httpclient"
	"gopkg.in/rightscale/rsc.v9/log"

	// phoney reference to make Godep pull this in for the code generators
	_ "github.com/go-openapi/inflect"
)

// Command line client entry point.
func main() {
	app := kingpin.New("rsc", "A RightScale API client")
	app.Writer(os.Stdout)
	app.Version(VV)

	cmdLine, err := ParseCommandLine(app)
	if err != nil {
		line := strings.Join(os.Args, " ")
		PrintFatal("%s: %s", line, err.Error())
	}

	resp, err := ExecuteCommand(app, cmdLine)
	if err != nil {
		PrintFatal("%s", err.Error())
	}
	if resp == nil {
		return // No results, just exit (e.g. setup, printed help...)
	}

	var notExactlyOneError bool
	displayer, err := NewDisplayer(resp)
	if err != nil {
		PrintFatal("%s", err.Error())
	}
	if resp.StatusCode < 200 || resp.StatusCode > 299 {
		// Let user know if something went wrong
		fmt.Fprintln(errOut, resp.Status)
		if len(displayer.body) > 0 {
			fmt.Fprintln(errOut, displayer.body)
		}
	} else if cmdLine.ExtractOneSelect != "" {
		err = displayer.ApplySingleExtract(cmdLine.ExtractOneSelect)
		if err != nil {
			notExactlyOneError = strings.Contains(err.Error(),
				"instead of one value") // Ugh, there has to be a better way
			PrintError(err.Error())
		}
		fmt.Fprint(out, displayer.Output())
	} else {
		if cmdLine.ExtractSelector != "" {
			err = displayer.ApplyExtract(cmdLine.ExtractSelector, false)
		} else if cmdLine.ExtractSelectorJSON != "" {
			err = displayer.ApplyExtract(cmdLine.ExtractSelectorJSON, true)
		} else if cmdLine.ExtractHeader != "" {
			err = displayer.ApplyHeaderExtract(cmdLine.ExtractHeader)
		}
		if err != nil {
			PrintFatal("%s", err.Error())
		} else if cmdLine.Pretty {
			displayer.Pretty()
		}
		fmt.Fprint(out, displayer.Output())
	}
	// Figure out exit code
	exitStatus := 0
	switch {
	case notExactlyOneError:
		exitStatus = 6
	case resp.StatusCode == 401:
		exitStatus = 1
	case resp.StatusCode == 403:
		exitStatus = 3
	case resp.StatusCode == 404:
		exitStatus = 4
	case resp.StatusCode > 399 && resp.StatusCode < 500:
		exitStatus = 2
	case resp.StatusCode > 499:
		exitStatus = 5
	}
	//fmt.Fprintf(os.Stderr, "exitStatus=%d\n", exitStatus)
	osExit(exitStatus)
}

// Helper that runs command line with give command client
func runCommand(client cmd.CommandClient, cmdLine *cmd.CommandLine) (resp *http.Response, err error) {
	cmds := strings.Split(cmdLine.Command, " ")
	if cmdLine.ShowHelp {
		err = client.ShowCommandHelp(cmdLine.Command)
	} else if len(cmds) > 1 && cmds[1] == "actions" {
		err = client.ShowAPIActions(cmdLine.Command)
	} else {
		resp, err = client.RunCommand(cmdLine.Command)
	}
	return
}

func ExecuteCommand(app *kingpin.Application, cmdLine *cmd.CommandLine) (resp *http.Response, err error) {
	app.Writer(errOut)
	log.Interactive()
	topCommand := strings.Split(cmdLine.Command, " ")[0]
	switch topCommand {
	case "setup":
		err = CreateConfig(cmdLine.ConfigPath)
	case "json":
		var b []byte
		b, err = ioutil.ReadAll(os.Stdin)
		if err == nil {
			resp = CreateJSONResponse(b)
		}
	default:
		// retry any failed API response as specified by the retry flag
		for i := 0; i < cmdLine.Retry+1; i++ {
			resp, err = doAPIRequest(topCommand, cmdLine)
			if !shouldRetry(resp, err) {
				break
			}
		}
	}

	return resp, err
}

// Constructs an http response from JSON input from Stdin
func CreateJSONResponse(b []byte) (resp *http.Response) {
	// Remove UTF-8 Byte Order Mark if it exists
	b = bytes.TrimPrefix(b, []byte{0xef, 0xbb, 0xbf})
	resp = &http.Response{
		StatusCode: 200,
		Body:       ioutil.NopCloser(bytes.NewBuffer(b)),
	}
	return resp
}

func shouldRetry(resp *http.Response, err error) bool {
	if err != nil {
		if neterr, ok := err.(net.Error); ok && neterr.Timeout() {
			return true
		}
	}
	if resp != nil {
		if resp.StatusCode == 500 || resp.StatusCode == 503 {
			return true
		}
	}
	return false
}

var doAPIRequest = func(command string, cmdLine *cmd.CommandLine) (resp *http.Response, err error) {
	httpclient.ResponseHeaderTimeout = time.Duration(cmdLine.Timeout) * time.Second
	client, err := APIClient(command, cmdLine)
	if err == nil {
		resp, err = runCommand(client, cmdLine)
		if err == nil {
			return resp, err
		}
	}
	return nil, err
}
